### ISSUE TITLE
InvalidTickCount Error Leads to Four Skipped slots

### ISSUE THREAD

**Issue by brianlong** [solana-labs/solana#11653]
#### Problem
Whenever I see the InvalidTickCount error in my logs, the current slot is spoiled along with the three following slots. In this sequence, the Testnet leader made the 1st block, messed-up the 2nd block, and didn't submit the 3rd or 4th.

```
{"leader":"BSF2yD9mqzaixDaLEraF1en82EWaXx7wbaCqSuKppqG5","skipped":false,"slot":30817244},
{"leader":"BSF2yD9mqzaixDaLEraF1en82EWaXx7wbaCqSuKppqG5","skipped":true,"slot":30817245},
{"leader":"BSF2yD9mqzaixDaLEraF1en82EWaXx7wbaCqSuKppqG5","skipped":true,"slot":30817246},
{"leader":"BSF2yD9mqzaixDaLEraF1en82EWaXx7wbaCqSuKppqG5","skipped":true,"slot":30817247}
```

(30817246 & 30817247 do not appear in my logs)

The next leader in line skipped 30817248 and then submitted blocks in the next 3 slots.

```
{"leader":"Db2V7nPHc4sPHne87nYXPGn8Kv8rMsiWCAjgAXmpqcpC","skipped":true,"slot":30817248},
{"leader":"Db2V7nPHc4sPHne87nYXPGn8Kv8rMsiWCAjgAXmpqcpC","skipped":false,"slot":30817249},
{"leader":"Db2V7nPHc4sPHne87nYXPGn8Kv8rMsiWCAjgAXmpqcpC","skipped":false,"slot":30817250},
{"leader":"Db2V7nPHc4sPHne87nYXPGn8Kv8rMsiWCAjgAXmpqcpC","skipped":false,"slot":30817251}
```

In my stats, I see some validators who have a higher than average skip rate AND the skip rate of the leaders who immediately follow them is also higher. If we can identify the cause of this problem, we can lower the overall skip rate for the cluster.

#### Proposed Solution

Identify the root cause of the problem. Also, is there a way to avoid skipping the next three slots whenever we get one of these errors?

**Comment by brianlong** [comment]
Hypothesis: The InvalidTickCount error :point_up_2: is not uniformly distributed amongst validators. Some validators have a higher than average error rate and can cause problems for the leaders who follow them too.

**Comment by carllin** [comment]
@brianlong thanks for reporting!

I believe that the `InvalidTickCount` may be more of a symptom of forking/validators being unable to keep up with the network, rather than the root cause of the problem.

Let's first take a look at what causes the `InvalidTickCount` errors. In normal operation, essentially what this error is reporting is that a leader's broadcast of his slot has been interrupted. This occurs when in the middle of the leader's broadcast, they detecct another *heavier* fork, and reset to that fork instead. For example:

```
  /------1 ---- 2 -----3 -----4       
0
 \-------5-----6---7----8-----9
```

In the example above, say validator `A` is scheduled to lead for the four blocks `1-4`, and for blocks `9-13`

Say this leader was broadcasting for  block `2`, when they then detects a set of heavier blocks at `5`, he will drop his blocks at `2` and reset his `PoH` to `5`. This will lead to him dropping all `4` of his blocks because this leader is essentially resetting to the block boundary of of the beginning of another leader's blocks. Then when he starts his next block at `9`, broadcast stage will detect his last broadcasted block at `2` was interrupted and send out a shred to the network marking this slot `2` as *complete*, even though it's missing ticks. Other validators will then mark this slot as errored with `InvalidTickCount`

**Comment by brianlong** [comment]
Hi @carllin How does that affect the next leader? When I was manually investigating logs, the four skipped slots were almost always shared between the leader and the next in line. In my stats, it seems like certain nodes always seem to have a higher than average "Skipped after %". I think that maybe due to this InvalidTickCount issue.

**Comment by brianlong** [comment]
In the example that I posted above, the leader did not dump all 4 of their own blocks. They submitted the first and dropped the remaining three. Is there something that would then force the next leader to drop the 1st slot?

**Comment by carllin** [comment]
@brianlong I believe it could be the "MAX_GRACE_SLOTS" defined by this constant here: https://github.com/solana-labs/solana/blob/master/core/src/poh_recorder.rs#L32

What this essentially does is the next leader agrees to give the previous leader a buffer/grace period off `MAX_GRACE_SLOTS`, before submitting a competing block of virtual ticks that runs over the previous leader.

That means if the previous leader is a bit late (because of latency/they're on a different fork), then the next leader will give up a block before generating a competing block that runs over the previous leader

**Comment by carllin** [comment]
@brianlong here was an issue tracking this exact problem: https://github.com/solana-labs/solana/issues/7588

It seems like the PR that closed that issue removes grace ticks, only if the database saw no shreds from the previous leader: https://github.com/solana-labs/solana/pull/7764/files#diff-224747fdb593a07e1c30e3b2559081f7R142

However it does not check that those shreds from the previous leader were on the same fork as the one the next leader wants to generate on, so that's a possible change that may improve this situation 😄

**Comment by brianlong** [comment]
@carllin Anything that could prevent the error from cascading to the next leader will be a big improvement! Isolating the error should help with the 'Skipped % after leader' problem that I have been watching. It will also lower the cluster skip rate in general. I suspect that there is a vicious cycle between skipped slots & the # of forks. Reducing skipped slots by up to 25% could be huge!

I will try to track the InvalidTickCount error to identify nodes with abnormally high error rates. In those cases, the node operator will need to look for problems with their setup.

What is your process from here? Leave the issue open until someone picks it up for development?

**Comment by carllin** [comment]
@brianlong I've linked the proposed fix 😄

**Comment by mvines** [comment]
I think we can call this issue fixed by #11679
